File: 20_valid_parentheses/20_valid_parentheses.ts
function isValid(s: string): boolean {
  const hashParentheses: Map<string, string> = new Map([
    ["(", ")"],
    ["{", "}"],
    ["[", "]"],
  ]);
  let stackParentheses: any[] = [];
  for (const chr of s) {
    if (hashParentheses.has(chr)) {
      stackParentheses.push(hashParentheses.get(chr));
    } else if (stackParentheses.length === 0 || stackParentheses.pop() != chr) {
      return false;
    }
  }
  return stackParentheses.length === 0;
}

File: 20_valid_parentheses/20_valid_parentheses.cpp
class Solution
{
public:
	bool isValid(string s)
	{

		map<char, char> hashParentheses = {{'(', ')'}, {'[', ']'}, {'{', '}'}};
		stack<char> stackParentheses;

		for (auto ch : s)
		{
			if (hashParentheses.count(ch))
			{
				stackParentheses.push(hashParentheses[ch]);
			}
			else if (stackParentheses.empty() || stackParentheses.top() != ch)
			{
				return false;
			}
			else
			{
				stackParentheses.pop();
			}
		}
		return stackParentheses.empty();
	}
};
